### ISSUE TITLE
[ENHANCEMENT] Support updating VM networks and vlanconfigs

### ISSUE THREAD

**Issue by yaocw2020** [harvester/harvester#3372]
**Is your feature request related to a problem? Please describe.**

- As for v1.1.1, Harvester doesn't support updating VM networks. If users want to change the VM network, they need to delete it and create a new one. However, the old VM network usually has been used by some VMs and could not be deleted. 
- As for v1.1.1, it's not allowed to change the cluster network of vlanconfigs. In some cases, our customs wants to migrate vlanconfigs into a new cluster network. We'd better support cluster network migration for vlanconfigs.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by yaocw2020** [comment]
Test cases:
Case 1:
- Create vlanconfig
- Migrate vlanconfig to another cluster network

Case 2:
- Create vlanconfig, VM network and VM
- Migrate vlanconfig to another cluster network

Case 3:
- Create vlanconfig and VM network
- Update VM network 
    - VLAN ID
    - Cluster network
    - Type

Case 4:
- Create vlanconfig, VM network and VM
- Migrate vlanconfig. It would fail.
- Stop VM and migrate vlanconfig to make the VM network unready
- Restart VM. It would fail.
- Modify VM network to new cluster network and restart VM

Case 5:
- Crate vlanconfig, VM network and VM
- Update VM network. It would fail.
- Stop VM and update VM network 
- Restart VM

**Comment by TachunLin** [comment]
Verified **failed** on `master-676d1eca-head` (02/16). Move to implement stage.

### Result
In test case 4, when we stop vm -> migrate vlanconfig to another cluster network -> Edit vlan network to correct vlanconfig. 

After we start VM, the vm keep in `stopping` with error message 
  ```
  0/4 nodes are available: 4 node(s) didn't match Pod's node affinity/selector. preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
  ```
  ![image](https://user-images.githubusercontent.com/29251855/219592509-02fe8496-44a0-48e1-a127-936285fd6085.png)

Check the node scheduling of vm, we can find both vlanconfig exists
![image](https://user-images.githubusercontent.com/29251855/219592906-ee782b2d-1d5a-4547-8a62-59f2f7e2ea18.png)


### Test Information
* Test Environment: 4 nodes Harvester bare machines
* Harvester version: `master-676d1eca-head` (02/16)

### Verify Steps
1. Create cluster network 1: `cn1`
2. Create cluster network 2: `cn2` 
3. Create vlanconfig `nc1` on `cn1` and set `enps1f1` of uplink nic
4. Create vlan network 1216 
5. Create a VM with vlan network 1216
6. Shutdown VM 
7. Migrate vlan config `nc1` to clusterwork `cn2` 
8. Edit the vlan 1216, set the vlan config to `cn2` cluster network
9. Start VM

**Comment by yaocw2020** [comment]
Found a bug in the self-testing.
- Create a new cluster network
- Create a vlanconfig that covers all nodes under the cluster network
- Change the vlanconfig node selector to one specified node. We will see the error message below.
<img width="1554" alt="image" src="https://user-images.githubusercontent.com/7421463/220581516-8b31ebe8-85bf-4b37-8fc9-4e499183156b.png">

@TachunLin Please take a look before the regression testing. Thanks.

**Comment by guangbochen** [comment]
This is addressed via https://github.com/harvester/harvester/pull/3511 and is ready for testing with the master-head iso, thanks.

**Comment by TachunLin** [comment]
### Test Scenarios:

#### Case 1:  Migrate vlanconfig to another cluster network and vice versa

1. Create cluster network `cn1` and `cn2`
2. Create vlanconfig `nc1` on cluster network `cn1` and set uplink NIC to `enp179s0f0` on `node1`
3. Create vlanconfig `nc2` on cluster network `cn2` and set uplink NIC to `enp179s0f0` on `node2`
4. Migrate vlanconfig `nc1` to `cn2` 
5. Migrate vlanconfig `nc2` to `cn1` 
6. Migrate vlanconfig `nc1` back to `cn1`
7. Delete vlanconfig `nc2`
8. Delete cluster network `cn2`

#### Case 2: Prevent to migrate vlanconfig if already set on a Running VM 
1. Create vlanconfig `nc1` 
2. Create a VM network `vlan101` and set manual route CIDR and Gateway
3. Create a VM `vm101` and set network to `vlan101`
4. Migrate vlanconfig to another cluster network `cn2`
5. It would prompt error message

#### Case 3: Migrate vlanconfig already set on a running VM
1. Create cluster network `cn1` and `cn2`
2. Create vlanconfig `nc1`, 
3. Create VM network `vlan101` on vlanconfig `nc1`
4. Create a VM `vm101` and set vlan network `vlan101`
5. Migrate vlanconfig to another cluster network `cn2`. It would fail.
6. Stop VM 
7. Migrate vlanconfig, now the the VM network display `unready`
8. Restart VM. It would fail.
9. Modify VM network to new cluster network
10. Start VM it should be running correctly

#### Case 4: Update vlanconfig from all nodes to specific node and one node to another node

ref: https://github.com/harvester/harvester/issues/3372#issuecomment-1439707225

1. Create a new cluster network `cn1`
2. Create a vlanconfig `nc1` that covers all nodes under the cluster network
3. Change the vlanconfig node selector of `nc1` to one specified `node 1`
4. We should change vlanconfig by using node selector 
5. Switch node selector of `nc1` to cover all nodes
6. Change node selector to `node 0` again


#### Case 5: Update VM network of the running VM
Prerequisite: 
  - Two active vlan network environment
  - Two cluster networks `cn1` and `cn2`


1. Create vlanconfig `nc1` on cluster network `cn1`
2. Create vlanconfig `nc2` on cluster network `cn2`
3. Create a VM network `vlan101` and set manual route CIDR and Gateway
4. Create a VM `vm101` and set network to `vlan101`
5. Update VM network
    - Change VLAN ID to `102` , it wil prompt error to prevent to proceed
    - Change Cluster network to `cn2`
    - Change VLAN manual route to correct CIDR and Gateway
    - Change Type to `UntaggedNetwork`

#### Case 6: Update VM network of the stopped VM
1. Follow Test Case 5 step 1 - 4
2. Stop the running VM
3. Stop VM
4. Update VM network
    - Change VLAN ID to `102` 
    - Change Cluster network to `cn2`
    - Change VLAN manual route to correct CIDR and Gateway
5. Start VM, check state is `Running`
6. Stop VM
7.  Update VM network Type to `UntaggedNetwork`
8. Start VM, check state is `Running`

**Comment by TachunLin** [comment]
Verified **failed** on `master-1222750f-head` (04/25). Move to implement stage. 

### Result
Encounter failure while validating the test scenario 1: 
1. We have two cluster network `cn1` and `cn2`, and two network config `nc1` and `nc2` set on each network
2. For each network config we set different uplink NICs and set node selector to all nodes 

When we migrate network config `nc1` to `nc2`, it will display error message
  ```
  admission webhook "validator.harvester-system.harvester-network-webhook" denied the request: Internal error
  occurred: could not update vlanConfig nc1 because it overlaps with other vlanconfigs matching node(s) 
  [node2 node3 node1]
  ```
Detailed steps can refer to the following video clip:

https://user-images.githubusercontent.com/29251855/234286263-82e74faf-519a-421a-8022-18c630b4eb4b.mp4



### Test Information
* Test Environment: 3 nodes harvester on bare machines
* Harvester version: `master-1222750f-head` 

### Verify Steps

1. Create cluster network cn1 and cn2
2. Create vlanconfig nc1 on cluster network cn1, set uplink NIC to `enp179s0f0` and select `all nodes`
3. Create vlanconfig nc2 on cluster network cn2, set uplink NIC to `enp179s0f1` and select `all nodes`
4. Migrate vlanconfig nc1 to cn2

**Comment by bk201** [comment]
@yaocw2020 Could you check David's test result?

**Comment by yaocw2020** [comment]
@TachunLin Since the network config nc1 under cn1 cluster network and network config nc2 under cn2 have selected all the three nodes, if you migrate nc1 from cn1 to cn2, it'll overlap with nc2.

**Comment by TachunLin** [comment]
@yaocw2020 , thank you for the explanation. 

According to our discussion and based on the [relationship between cluster network and network config](https://docs.harvesterhci.io/v1.1/networking/clusternetwork#relationship-between-cluster-network-network-config-vm-network) diagram

![image](https://github.com/harvester/harvester/assets/29251855/007af3d2-2d24-4d86-b116-bb4c4478d4f3)

The network config migration is based on the cluster network perspective and applied to nodes machine as the base unit. 

For the case I mentioned in https://github.com/harvester/harvester/issues/3372#issuecomment-1521768343, it would be expected to prompt the `overlaps with other vlanconfigs` failure since we can't migrate network config to another cluster network which already have overlap nodes applied on it. 

Instead if we want to achieve the test scenario, we could use the following steps
1. Delete the `nc1` network config on `cn1` to free up nic `enp179s0f0` 
2. Create a new network config `nc3` on cluster network `cn2` and select `enp179s0f0` as the uplink nic
3. Or edit the network config `nc2` to replace or add `enp179s0f0`

Just like to confirm whether my understanding is correct.

**Comment by yaocw2020** [comment]
Thanks @TachunLin, good understanding.

**Comment by TachunLin** [comment]
Verified **fixed** on `master-d739c2a4-head` (5/18). Close this issue.

### Result
We can pass all test scenarios: 
1. Case 1: Migrate vlanconfig to another cluster network and vice versa
  - can migrate network config between cluster network with non-overlap nodes
2. Case 2: Prevent to migrate vlanconfig if already set on a Running VM
  ![image](https://github.com/harvester/harvester/assets/29251855/05e75597-f2ee-41f0-b04f-e9b28b2313d3)
3. Case 3: Migrate vlanconfig already set on a running VM
  ![image](https://github.com/harvester/harvester/assets/29251855/498ea58b-45fa-460c-8bf5-878a9bf2048b)
4. Case 4: Update vlanconfig from all nodes to specific node and one node to another node
5. Case 5: Update VM network of the running VM
![image](https://github.com/harvester/harvester/assets/29251855/6d6095b1-f749-48b8-9541-22fa5582081d)
6. Update VM network of the stopped VM


### Test Information
* Test Environment: 3 nodes harvester on bare machines and kvm
* Harvester version: `master-d739c2a4-head` (5/18)
* Number to NICs on each node: 3

### Verify Steps
https://github.com/harvester/harvester/issues/3372#issuecomment-1521502390
